### PR TITLE
Add pseudo-index for key headers by height

### PR DESCRIPTION
**Pull request by cytadela8**
- Add {key_header, Height, Hash} -> Header to aec_chain_state database
- Proper migration
- Use new fast find key header by height when migration done

TODO:
- [ ] performance tests

### CODE REVIEW

File: apps/aecore/src/aec_db.erl [hunk 1]
```diff
@@ -351,7 +351,10 @@ write_block(Block, Hash) ->
             Headers = #aec_headers{ key = Hash
                                   , value = Header
                                   , height = Height },
-            ?t(mnesia:write(Headers),
+            ?t(begin
+                   mnesia:write(Headers),
+                   mnesia:write(#aec_chain_state{key = {key_block, Height, Hash}, value = Header})
```
Comment (gorbak25) [inline apps/aecore/src/aec_db.erl:356]: key_block -> key_header

File: apps/aecore/src/aec_chain_state.erl [hunk 2]
```diff
@@ -1346,3 +1348,78 @@ chain_ends_maybe_apply([H1|T1] = OldTops, [H2|T2], NewTops) -> %% Check if H1 is
...
+                            ok
+                    end;
+                in_progress -> %% The migration was interrupted and not finished yet
+                    lager:info("[Orphan key blocks scan] Resuming interrupted scan"),
+                    start_key_headers_height_store_migration()
+            end
+    end.
+
+start_key_headers_height_store_migration() ->
+    lager:info("[Key headers migrations scan] Retriving all key blocks"),
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain_state.erl:1376]: key headers*

**Comment by gorbak25** [comment]
@cytadela8 @dincho  Just sacrificed 18.218.165.37 for testing out the migration :)
```
2021-04-01 21:03:56.405 [info] <0.1529.0>@aec_chain_state:ensure_key_headers_height_store:1370 [Orphan key blocks scan] Resuming interrupted scan
2021-04-01 21:03:56.405 [info] <0.1529.0>@aec_chain_state:start_key_headers_height_store_migration:1376 [Key headers migrations scan] Retriving all key blocks
2021-04-01 21:04:14.981 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 18575295 microseconds
2021-04-01 21:04:15.689 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 707995 microseconds
2021-04-01 21:04:25.828 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 10139008 microseconds
2021-04-01 21:04:26.393 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 565203 microseconds
2021-04-01 21:04:36.399 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 10005600 microseconds
2021-04-01 21:04:37.053 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 654559 microseconds
2021-04-01 21:04:44.952 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7898636 microseconds
2021-04-01 21:04:45.373 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 420921 microseconds
2021-04-01 21:04:52.496 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7122287 microseconds
2021-04-01 21:04:52.925 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 429493 microseconds
2021-04-01 21:05:00.232 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7306992 microseconds
2021-04-01 21:05:00.671 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 438265 microseconds
2021-04-01 21:05:10.032 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 9361321 microseconds
2021-04-01 21:05:10.563 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 530932 microseconds
2021-04-01 21:05:17.949 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7386160 microseconds
2021-04-01 21:05:18.390 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 440703 microseconds
2021-04-01 21:05:25.455 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7064382 microseconds
2021-04-01 21:05:25.832 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 377188 microseconds
2021-04-01 21:05:33.150 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7317921 microseconds
2021-04-01 21:05:33.584 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 433693 microseconds
2021-04-01 21:05:43.030 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 9446087 microseconds
2021-04-01 21:05:43.652 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 622295 microseconds
2021-04-01 21:05:50.936 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7284113 microseconds
2021-04-01 21:05:51.366 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 429818 microseconds
2021-04-01 21:05:59.061 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7694639 microseconds
2021-04-01 21:05:59.563 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 501657 microseconds
2021-04-01 21:06:06.837 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 7274611 microseconds
2021-04-01 21:06:07.258 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 420752 microseconds
2021-04-01 21:06:16.838 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 9579584 microseconds
2021-04-01 21:06:17.280 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 442278 microseconds
2021-04-01 21:06:25.358 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 8077343 microseconds
2021-04-01 21:06:25.777 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 419532 microseconds
2021-04-01 21:06:37.201 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11423210 microseconds
2021-04-01 21:06:37.691 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 489888 microseconds
2021-04-01 21:06:48.748 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11057374 microseconds
2021-04-01 21:06:49.167 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 419113 microseconds
2021-04-01 21:07:02.393 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 13225996 microseconds
2021-04-01 21:07:02.829 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 435456 microseconds
2021-04-01 21:07:14.240 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11411392 microseconds
2021-04-01 21:07:14.660 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 419818 microseconds
2021-04-01 21:07:26.149 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11488377 microseconds
2021-04-01 21:07:26.591 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 441929 microseconds
2021-04-01 21:07:37.891 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11300513 microseconds
2021-04-01 21:07:38.314 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 421945 microseconds
2021-04-01 21:07:51.934 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 13620683 microseconds
2021-04-01 21:07:52.318 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 383226 microseconds
2021-04-01 21:08:03.660 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11342784 microseconds
2021-04-01 21:08:04.154 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 493207 microseconds
2021-04-01 21:08:15.671 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11517449 microseconds
2021-04-01 21:08:16.111 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 439344 microseconds
2021-04-01 21:08:27.563 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11452303 microseconds
2021-04-01 21:08:27.948 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 385228 microseconds
2021-04-01 21:08:42.130 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 14181532 microseconds
2021-04-01 21:08:42.560 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 429821 microseconds
2021-04-01 21:08:54.678 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 12117422 microseconds
2021-04-01 21:08:55.116 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 438630 microseconds
2021-04-01 21:09:07.114 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11997322 microseconds
2021-04-01 21:09:07.595 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 481654 microseconds
2021-04-01 21:09:19.250 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11654376 microseconds
2021-04-01 21:09:19.675 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 424585 microseconds
2021-04-01 21:09:34.726 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 15051506 microseconds
2021-04-01 21:09:35.273 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 546343 microseconds
2021-04-01 21:09:47.592 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 12319680 microseconds
2021-04-01 21:09:48.060 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 467195 microseconds
2021-04-01 21:09:59.340 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11280114 microseconds
2021-04-01 21:09:59.785 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 444931 microseconds
2021-04-01 21:10:11.167 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11382270 microseconds
2021-04-01 21:10:11.578 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 410869 microseconds
2021-04-01 21:10:25.518 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 13939498 microseconds
2021-04-01 21:10:26.083 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 565595 microseconds
2021-04-01 21:10:39.052 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 12968238 microseconds
2021-04-01 21:10:39.500 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 447761 microseconds
2021-04-01 21:10:50.678 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 11178112 microseconds
2021-04-01 21:10:51.080 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 402168 microseconds
2021-04-01 21:11:01.968 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 10888243 microseconds
2021-04-01 21:11:02.380 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 411720 microseconds
2021-04-01 21:11:16.119 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 13738261 microseconds
2021-04-01 21:11:16.713 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 594853 microseconds
2021-04-01 21:11:29.448 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 12733997 microseconds
2021-04-01 21:11:29.886 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 10000 headers in 438064 microseconds
2021-04-01 21:11:39.551 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1407 [Key headers migrations scan] Read headers in 9665649 microseconds
2021-04-01 21:11:39.858 [info] <0.1579.0>@aec_chain_state:key_headers_height_store_migration_step:1420 [Key headers migrations scan] Wrote 8274 headers in 306160 microseconds
2021-04-01 21:11:39.858 [info] <0.1579.0>@aec_chain_state:start_key_headers_height_store_migration:1391 [Key headers migrations scan] DONE: In total migrated 408274 headers in 463445068 microseconds
```
Looks good :) I will run some more tests on that sacrificed node.

**Comment by gorbak25** [comment]
@cytadela8 @hanssv 
image
300 microseconds is a nice improvement from 10-20ms we had previously :)
Let me check whether we can make it faster.

**Comment by gorbak25** [comment]
```
(aeternity@localhost)1> {rocksdb, #{aec_chain_state := H}} = persistent_term:get({aec_db,mnesia_bypass}).
{rocksdb,#{aec_account_state =>
               #Ref<0.1224451220.928120841.466>,
           aec_block_state => #Ref<0.1224451220.928120841.470>,
           aec_blocks => #Ref<0.1224451220.927858693.261492>,
           aec_call_state => #Ref<0.1224451220.927858693.261494>,
           aec_chain_state => #Ref<0.1224451220.928120841.472>,
           aec_channel_state => #Ref<0.1224451220.928120841.474>,
           aec_contract_state => #Ref<0.1224451220.927858693.261496>,
           aec_discovered_pof => #Ref<0.1224451220.927858693.261498>,
           aec_headers => #Ref<0.1224451220.928120841.464>,
           aec_headers_index => #Ref<0.1224451220.928120841.468>,
           aec_name_service_cache => #Ref<0.1224451220.928120841.476>,
           aec_name_service_state =>
               #Ref<0.1224451220.927858693.261500>,
           aec_oracle_cache => #Ref<0.1224451220.927858693.261502>,
           aec_oracle_state => #Ref<0.1224451220.928120841.478>,
           aec_peers => #Ref<0.1224451220.927858693.261504>,
           aec_signal_count => #Ref<0.1224451220.927858693.261506>,
           aec_signed_tx => #Ref<0.1224451220.927858693.261508>,
           aec_tx_location => #Ref<0.1224451220.927858693.261510>,
           aec_tx_pool => #Ref<0.1224451220.927858693.261512>}}
(aeternity@localhost)2> WI = fun(Height, F) -> {ok, I} = rocksdb:iterator(H, [{iterate_upper_bound, sext:prefix({key_block, Height+1, '_'})}]), try F(I) after rocksdb:iterator_close(I) end end.
#Fun<erl_eval.13.126501267>
(aeternity@localhost)3> G = fun G(I, {ok, K ,V}, Height, Acc) -> case sext:decode(K) of {_, Height, Hash} -> G(I, rocksdb:iterator_move(I, next), Height, [{Hash, aec_headers:from_db_header(element(3, binary_to_term(V)))}|Acc]); _ -> Acc end; G(I, _, _, Acc) -> Acc end.
#Fun<erl_eval.29.126501267>
(aeternity@localhost)4> F = fun(Height) -> WI(Height, fun(I) -> 
(aeternity@localhost)4> G(I, rocksdb:iterator_move(I, sext:prefix({key_block, Height, '_'})), Height, [])
(aeternity@localhost)4> end) end.
#Fun<erl_eval.7.126501267>
(aeternity@localhost)5> {T, _} = timer:tc(fun() -> [F(X) || X <- lists:seq(0, 500000)] end), ok.
ok
(aeternity@localhost)6> T/500000.
409.6429
(aeternity@localhost)7> 
```
@cytadela8 Could you push a version of your PR with the above bypass logic? I don't want to compare interpreted code vs compiled code.

**Comment by cytadela8** [comment]
Generally it would be nice to have some actual data where extacly we spend how much time.

**Comment by gorbak25** [comment]
Initial benchmarks for #3578 show that using rocksdb directly is actually slower than this approach - rerunning my benchmark.

**Comment by gorbak25** [comment]
Hmm second run yields ~200 us per operation

**Comment by gorbak25** [comment]
image
@cytadela8 looks like the dirty version is actually faster.

**Comment by gorbak25** [comment]
@cytadela8 include the cleaned up version dirty-dirty version in this PR.

**Comment by cytadela8** [comment]
Replaced by #3578

**Pull request closed without merging**